File: ngraph/src/ngraph/op/cum_sum.cpp
//*****************************************************************************
// Copyright 2017-2020 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//*****************************************************************************

#include "ngraph/op/cum_sum.hpp"
#include "ngraph/attribute_visitor.hpp"
#include "ngraph/graph_util.hpp"
#include "ngraph/op/broadcast.hpp"
#include "ngraph/op/constant.hpp"
#include "ngraph/runtime/reference/cum_sum.hpp"

using namespace std;
using namespace ngraph;

constexpr NodeTypeInfo op::v0::CumSum::type_info;

op::v0::CumSum::CumSum(const Output<Node>& arg,
                       const Output<Node>& axis,
                       const bool exclusive,
                       const bool reverse)
    : Op({arg, axis})
    , m_exclusive(exclusive)
    , m_reverse(reverse)
{
    constructor_validate_and_infer_types();
}

op::v0::CumSum::CumSum(const Output<Node>& arg, const bool exclusive, const bool reverse)
    : Op({arg, op::Constant::create(element::i32, Shape{}, {0})})
    , m_exclusive(exclusive)
    , m_reverse(reverse)
{
    constructor_validate_and_infer_types();
}

bool op::v0::CumSum::visit_attributes(AttributeVisitor& visitor)
{
    visitor.on_attribute("exclusive", m_exclusive);
    visitor.on_attribute("reverse", m_reverse);
    return true;
}

void op::v0::CumSum::validate_and_infer_types()
{
    element::Type arg_type = get_input_element_type(0);
    PartialShape arg_shape = get_input_partial_shape(0);
    set_output_type(0, arg_type, arg_shape);

    PartialShape axes_shape{PartialShape::dynamic()};
    if (get_input_partial_shape(1).is_static())
    {
        axes_shape = get_input_partial_shape(1);
    }

    const auto& axis_type = get_input_element_type(1);
    NODE_VALIDATION_CHECK(this,
                          axis_type == element::i32 || axis_type == element::i64,
                          "axis element type must be either int64_t or int32_t but got (",
                          axis_type,
                          ").");
}

shared_ptr<Node> op::v0::CumSum::clone_with_new_inputs(const OutputVector& new_args) const
{
    check_new_args_count(this, new_args);
    return make_shared<op::CumSum>(new_args.at(0), new_args.at(1), m_exclusive, m_reverse);
}

void op::v0::CumSum::generate_adjoints(autodiff::Adjoints& adjoints, const OutputVector& deltas)
{
    auto delta = deltas.at(0);
    auto input_tensor = input_value(0);
    adjoints.add_delta(input_tensor, delta);
}

shared_ptr<Node> op::v0::CumSum::get_default_value() const
{
    return ngraph::make_constant_from_string("0", get_element_type(), get_shape());
}

namespace {
    template<element::Type_t ET>
    inline bool
    evaluate(const HostTensorVector &args, const HostTensorPtr &out,bool exclusive, bool reverse) {
        using T = typename element_type_traits<ET>::value_type;
        // TODO: For validation purposes only i64 axis_tensor is used. Types coverage have to be extended if needed
        using P = typename element_type_traits<ngraph::element::Type_t::i64>::value_type;
        runtime::reference::cumsum<T, P>(args[0]->get_data_ptr<ET>(),
                                         args[1]->get_data_ptr<ngraph::element::Type_t::i64>(),
                                         out->get_data_ptr<ET>(), args[0]->get_shape(),
                                         exclusive, reverse);
        return true;
    }

    bool evaluate_cumsum(const HostTensorVector &args, const HostTensorPtr &out,bool exclusive, bool reverse) {
        bool rc = true;

        switch (out->get_element_type()) {
            TYPE_CASE(i8)(args, out,exclusive, reverse);
                break;
            TYPE_CASE(i16)(args, out,exclusive, reverse);
                break;
            TYPE_CASE(i32)(args, out,exclusive, reverse);
                break;
            TYPE_CASE(u8)(args, out,exclusive, reverse);
                break;
            TYPE_CASE(f16)(args, out,exclusive, reverse);
                break;
            TYPE_CASE(f32)(args, out,exclusive, reverse);
                break;
            default:
                rc = false;
                break;
        }
        return rc;
    }
}

bool op::CumSum::evaluate(const HostTensorVector &outputs, const HostTensorVector &inputs) {
    return evaluate_cumsum(inputs, outputs[0], m_exclusive, m_reverse);
}

File: model-optimizer/extensions/ops/cumsum.py
"""
 Copyright (C) 2018-2021 Intel Corporation

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
"""
import numpy as np

from mo.front.extractor import bool_to_str
from mo.graph.graph import Node, Graph
from mo.ops.op import Op


def cumsum(a, axis=None, exclusive=False, reverse=False):
    if reverse:
        a = np.flip(a, axis)
    res = np.cumsum(a, axis=axis)
    if exclusive:
        res -= a
    if reverse:
        res = np.flip(res, axis)
    return res


class CumSum(Op):
    enabled = False
    op = 'CumSum'
    version = 'opset3'

    def __init__(self, graph: Graph, attrs: dict):
        super().__init__(graph, {
            'op': self.op,
            'type': self.op,
            'version': self.version,

            'infer': self.infer,

            'in_ports_count': 2,
            'out_ports_count': 1,
        }, attrs)

    def supported_attrs(self):
        return [('exclusive', lambda node: bool_to_str(node, 'exclusive')),
                ('reverse', lambda node: bool_to_str(node, 'reverse'))]

    @staticmethod
    def infer(node: Node):
        node_name = node.soft_get('name', node.id)

        input_shape = node.in_port(0).data.get_shape()
        assert input_shape is not None, 'Input shape is None for node "{}"'.format(node_name)
        if not node.in_port(1).disconnected():
            assert len(node.in_port(1).data.get_shape()) == 0, 'Axis is not scalar for node: {}'.format(node_name)

        node.out_port(0).data.set_shape(input_shape.copy())

        input_value = node.in_port(0).data.get_value()
        if input_value is not None:
            axis = None if node.in_port(1).disconnected() else node.in_port(1).data.get_value()
            reverse = node.reverse if node.has_valid('reverse') else False
            exclusive = node.exclusive if node.has_valid('exclusive') else False
            node.out_port(0).data.set_value(cumsum(input_value, axis=axis, reverse=reverse, exclusive=exclusive))


class MXNetCumSum(Op):
    enabled = False
    op = 'MXNetCumSum'

    def __init__(self, graph: Graph, attrs: dict):
        super().__init__(graph, {
            'op': self.op,
            'type': None,

            'infer': None,

            'in_ports_count': 1,
            'out_ports_count': 1,
        }, attrs)
